File: OneSignalSDK/onesignal/src/main/java/com/onesignal/outcomes/OSOutcomeEventsV1Repository.java
package com.onesignal.outcomes;

import com.onesignal.OSLogger;
import com.onesignal.OneSignalApiResponseHandler;
import com.onesignal.OutcomeEvent;
import com.onesignal.outcomes.domain.OutcomeEventsService;
import com.onesignal.outcomes.model.OSOutcomeEventParams;

import org.json.JSONException;
import org.json.JSONObject;

class OSOutcomeEventsV1Repository extends OSOutcomeEventsRepository {

    private static final String DIRECT = "direct";

    OSOutcomeEventsV1Repository(OSLogger logger, OSOutcomeEventsCache outcomeEventsCache, OutcomeEventsService outcomeEventsService) {
        super(logger, outcomeEventsCache, outcomeEventsService);
    }

    @Override
    public void requestMeasureOutcomeEvent(String appId, int deviceType, OSOutcomeEventParams eventParams, OneSignalApiResponseHandler responseHandler) {
        OutcomeEvent event = OutcomeEvent.fromOutcomeEventParamsV2toOutcomeEventV1(eventParams);
        switch (event.getSession()) {
            case DIRECT:
                requestMeasureDirectOutcomeEvent(appId, deviceType, event, responseHandler);
                break;
            case INDIRECT:
                requestMeasureIndirectOutcomeEvent(appId, deviceType, event, responseHandler);
                break;
            case UNATTRIBUTED:
                requestMeasureUnattributedOutcomeEvent(appId, deviceType, event, responseHandler);
                break;
            case DISABLED:
                // In this stage we should't have this case
        }
    }

    private void requestMeasureDirectOutcomeEvent(String appId, int deviceType, OutcomeEvent event, OneSignalApiResponseHandler responseHandler) {
        try {
            JSONObject jsonObject = event.toJSONObjectForMeasure();
            jsonObject.put(APP_ID, appId);
            jsonObject.put(DEVICE_TYPE, deviceType);
            jsonObject.put(DIRECT, true);

            outcomeEventsService.sendOutcomeEvent(jsonObject, responseHandler);
        } catch (JSONException e) {
            logger.error("Generating direct outcome:JSON Failed.", e);
        }
    }

    private void requestMeasureIndirectOutcomeEvent(String appId, int deviceType, OutcomeEvent event, OneSignalApiResponseHandler responseHandler) {
        try {
            JSONObject jsonObject = event.toJSONObjectForMeasure();
            jsonObject.put(APP_ID, appId);
            jsonObject.put(DEVICE_TYPE, deviceType);
            jsonObject.put(DIRECT, false);

            outcomeEventsService.sendOutcomeEvent(jsonObject, responseHandler);
        } catch (JSONException e) {
            logger.error("Generating indirect outcome:JSON Failed.", e);
        }
    }

    private void requestMeasureUnattributedOutcomeEvent(String appId, int deviceType, OutcomeEvent event, OneSignalApiResponseHandler responseHandler) {
        try {
            JSONObject jsonObject = event.toJSONObjectForMeasure();
            jsonObject.put(APP_ID, appId);
            jsonObject.put(DEVICE_TYPE, deviceType);

            outcomeEventsService.sendOutcomeEvent(jsonObject, responseHandler);
        } catch (JSONException e) {
            logger.error("Generating unattributed outcome:JSON Failed.", e);
        }
    }

}

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/outcomes/data/OSOutcomeEventsV1Repository.kt
package com.onesignal.outcomes.data

import com.onesignal.OSLogger
import com.onesignal.OneSignalApiResponseHandler
import com.onesignal.OSOutcomeEvent
import com.onesignal.influence.domain.OSInfluenceType
import com.onesignal.outcomes.OSOutcomeConstants
import com.onesignal.outcomes.domain.OSOutcomeEventParams
import org.json.JSONException

internal class OSOutcomeEventsV1Repository(logger: OSLogger,
                                           outcomeEventsCache: OSOutcomeEventsCache,
                                           outcomeEventsService: OutcomeEventsService) : OSOutcomeEventsRepository(logger, outcomeEventsCache, outcomeEventsService) {
    override fun requestMeasureOutcomeEvent(appId: String, deviceType: Int, eventParams: OSOutcomeEventParams, responseHandler: OneSignalApiResponseHandler) {
        val event = OSOutcomeEvent.fromOutcomeEventParamsV2toOutcomeEventV1(eventParams)
        when (event.session) {
            OSInfluenceType.DIRECT -> requestMeasureDirectOutcomeEvent(appId, deviceType, event, responseHandler)
            OSInfluenceType.INDIRECT -> requestMeasureIndirectOutcomeEvent(appId, deviceType, event, responseHandler)
            OSInfluenceType.UNATTRIBUTED -> requestMeasureUnattributedOutcomeEvent(appId, deviceType, event, responseHandler)
            else -> {
            }
        }
    }

    private fun requestMeasureDirectOutcomeEvent(appId: String, deviceType: Int, event: OSOutcomeEvent, responseHandler: OneSignalApiResponseHandler) {
        try {
            event.toJSONObjectForMeasure()
                    .put(OSOutcomeConstants.APP_ID, appId)
                    .put(OSOutcomeConstants.DEVICE_TYPE, deviceType)
                    .put(OSOutcomeConstants.DIRECT_PARAM, true)
                    .also { jsonObject ->
                        outcomeEventsService.sendOutcomeEvent(jsonObject, responseHandler)
                    }
        } catch (e: JSONException) {
            logger.error("Generating direct outcome:JSON Failed.", e)
        }
    }

    private fun requestMeasureIndirectOutcomeEvent(appId: String, deviceType: Int, event: OSOutcomeEvent, responseHandler: OneSignalApiResponseHandler) {
        try {
            event.toJSONObjectForMeasure()
                    .put(OSOutcomeConstants.APP_ID, appId)
                    .put(OSOutcomeConstants.DEVICE_TYPE, deviceType)
                    .put(OSOutcomeConstants.DIRECT_PARAM, false)
                    .also { jsonObject ->
                        outcomeEventsService.sendOutcomeEvent(jsonObject, responseHandler)
                    }
        } catch (e: JSONException) {
            logger.error("Generating indirect outcome:JSON Failed.", e)
        }
    }

    private fun requestMeasureUnattributedOutcomeEvent(appId: String, deviceType: Int, event: OSOutcomeEvent, responseHandler: OneSignalApiResponseHandler) {
        try {
            event.toJSONObjectForMeasure()
                    .put(OSOutcomeConstants.APP_ID, appId)
                    .put(OSOutcomeConstants.DEVICE_TYPE, deviceType)
                    .also { jsonObject ->
                        outcomeEventsService.sendOutcomeEvent(jsonObject, responseHandler)
                    }
        } catch (e: JSONException) {
            logger.error("Generating unattributed outcome:JSON Failed.", e)
        }
    }
}